op-supervisor/supervisor/backend/rewinder/rewinder.go

```diff
@@ -0,0 +1,256 @@
+	}
+
+	// If we're still on the canonical chain, nothing to do
+	if crossSafeL1.Hash == newTip.Hash {
+		return nil
+	}
+
+	// Get the finalized block as our lower bound
+	finalized, err := r.db.Finalized(chainID)
+	if err != nil {
+		return fmt.Errorf("failed to get finalized block for chain %s: %w", chainID, err)
+	}
+	finalizedL1, err := r.db.CrossDerivedFromBlockRef(chainID, finalized.ID())
+	if err != nil {
+		return fmt.Errorf("failed to get finalized L1 block for chain %s: %w", chainID, err)
+	}
+
+	// Find the common ancestor by walking back through L1 blocks
+	commonL1Ancestor := finalizedL1.ID()
+	currentL1 := crossSafeL1.ID()
+	for currentL1.Number >= finalizedL1.Number {
+		// Get the canonical L1 block at this height from the node
+		remoteL1, err := r.l1Node.L1BlockRefByNumber(context.Background(), currentL1.Number)
+		if err != nil {
+			return fmt.Errorf("failed to get L1 block at height %d: %w", currentL1.Number, err)
+		}
+
+		// If hashes match, we found the common ancestor
+		if remoteL1.Hash == currentL1.Hash {
+			commonL1Ancestor = currentL1
+			break
+		}
+
+		// Get the previous L1 block from our DB
+		prevDerivedFrom, err := r.db.PreviousDerivedFrom(chainID, currentL1)
+		if err != nil {
+			// If we hit the first block, use it as common ancestor
+			if errors.Is(err, types.ErrPreviousToFirst) {
+				// Still need to verify this block is canonical
+				remoteFirst, err := r.l1Node.L1BlockRefByNumber(context.Background(), currentL1.Number)
+				if err != nil {
+					return fmt.Errorf("failed to get first L1 block: %w", err)
+				}
+				if remoteFirst.Hash == currentL1.Hash {
+					commonL1Ancestor = currentL1
+				} else {
+					// First block isn't canonical, use finalized
+					commonL1Ancestor = finalizedL1.ID()
+				}
+				break
+			}
+			return fmt.Errorf("failed to get previous L1 block: %w", err)
+		}
+
+		// Move to the parent
+		currentL1 = prevDerivedFrom.ID()
+	}
+
+	// Get the last L2 block derived from the common ancestor
+	crossSafeDerived, err := r.db.LastDerivedFrom(chainID, commonL1Ancestor)
+	if err != nil {
+		return fmt.Errorf("failed to get derived from for chain %s: %w", chainID, err)
+	}
+
+	// Rewind CrossSafe to the derived block
+	if err := r.db.RewindCrossSafe(chainID, crossSafeDerived); err != nil {
+		return fmt.Errorf("failed to rewind cross-safe for chain %s: %w", chainID, err)
+	}
+
+	// Now check LocalSafe DB for reorg
+	localSafeDerived, err := r.checkLocalSafeForReorg(chainID, crossSafeDerived)
+	if err != nil {
+		return fmt.Errorf("failed to check local-safe for reorg for chain %s: %w", chainID, err)
+	}
+
+	// Rewind LocalSafe to the derived block
+	if err := r.db.RewindLocalSafe(chainID, localSafeDerived); err != nil {
+		return fmt.Errorf("failed to rewind local-safe for chain %s: %w", chainID, err)
+	}
+
+	// Rewind logs DB to match
+	if err := r.db.RewindLogs(chainID, localSafeDerived); err != nil {
+		return fmt.Errorf("failed to rewind logs for chain %s: %w", chainID, err)
+	}
```

Comment:
I think @protolambda 's suggestion here was to not rewind the Logs DB at all if there's been an L1 reorg. The reasoning being that when an L1 reorg occurs, it doesn't usually affect the L2's data. The batches might end up being published in different blocks or in different indexes, but the contents of the batches are the same, meaning the unsafe chain doesn't care about L1 reorgs.

Once new safe data is published *then* if the unsafe doesn't match, we can prune it (called consolidation in the op-node). But you have that path handled in the `handleDerivationEvent` function.